docs/architecture/adr-007-address-generation-of-chain-link-proof.md

```diff
@@ -0,0 +1,160 @@
+
+`Bech32Address` relies on the address generation algorithm specific of Cosmos-SDK chains:
+- for single signature accounts: 
+  ```
+  ripemd160(sha256(32 bytes public key))[:20]
+  ``` 
+- for multi-sig accounts: 
+  ```
+  sha256(aminoCdc.Marshal(multisig public key))[:20]
+  ```
+
+`Base58Address` relies on the Solana algorithm:
+```
+(32 bytes public key)[:32]
+```
+
+`HexAddress` relies on the Ethereum algorithm:
+```
+keccak64(64 bytes public key)[12:]
+```
+
+Due to the fact all the `AddressData` implementations are highly coupled to a single address generation function, it is currently impossible to link an address that needs to be encoded using a supported encoding method but is generated using a different algorithm. This is the case of Elrond addresses which are encoded using the `Bech32` encoding algorithm, but are generated using an algorithm that is different from the Cosmos one.
+
+## Decision
+
+In order to allow developers to integrate any kind of address, we will review how the current `Address` structure is made. Instead of using an interface with multiple implementations based on the encoding algorithm, we will use a single structure that allows to specify both the encoding algorithm and the hashing algorithm(s) that should be used to get the proper address: 
+
+```protobuf
+syntax = "proto3";
+
+// Address contains the data of an external address
+message Address {
+  // Encoded value of the address
+  string value = 1;
+  
+  // Algorithm that has been used in order to generate the address starting from the public key bytes
+  GenerationAlgorithm generation_algorithm = 2;
+  
+  // Algorithm that needs to be used to properly encode the address 
+  google.proto.Any encoding_algorithm = 3 [ (cosmos_proto.accepts_interface) = "AddressEncoding" ];
+}
+
+// GenerationAlgorithm represents various address generation algorithms
+enum GenerationAlgorithm {
+  // GENERATION_ALGORITHM_UNKNOWN represents an unknown algorithm and will be discarded 
+  GENERATION_ALGORITHM_UNKNOWN = 0;
+  
+  // GENERATION_ALGORITHM_COSMOS represents the Cosmos generation algorithm
+  GENERATION_ALGORITHM_COSMOS = 1;
+  
+  // GENERATION_ALGORITHM_EVM represents the EVM generation algorithm  
+  GENERATION_ALGORITHM_EVM = 2;
+  
+  // GENERATION_ALGORITHM_DO_NOTHING should be used when the public key bytes do not need to be modified  
+  GENERATION_ALGORITHM_DO_NOTHING = 3;
+}
+
+// Bech32Encoding represents the encoding algorithm based on the Bech32 format
+message Bech32Encoding {
+  option (cosmos_proto.implements_interface) = "AddressEncoding";
+  
+  // Prefix to be used
+  string prefix = 1; 
+}
+
+// Base58Encoding represents the encoding algorithm based on the Base58 format
+message Base58Encoding {
+  option (cosmos_proto.implements_interface) = "AddressEncoding";
+
+  // Prefix to be used
+  string prefix = 1;
```

Comment:
I think we should provide the prefix field since Polkadot base58 address format has prefix.
References:
https://wiki.polkadot.network/docs/learn-accounts#address-format

Reply:
I've nothing against this, since it brings more support for other chains.
Thoughts? @RiccardoM

Reply:
Nothing against, just made it optional though